File: project/drivers/driver/ruby/lib/scratchbird/metadata.rb
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

module Scratchbird
  module Metadata
    SCHEMAS_QUERY = "SELECT schema_id, schema_name, owner_id, default_tablespace_id FROM sys.schemas WHERE is_valid = 1 ORDER BY schema_name"
    TABLES_QUERY = "SELECT table_id, schema_id, table_name, table_type, owner_id FROM sys.tables WHERE is_valid = 1 ORDER BY table_name"
    COLUMNS_QUERY = "SELECT column_id, table_id, column_name, data_type_id, data_type_name, ordinal_position, is_nullable, default_value, domain_id, collation_id, charset_id, is_identity, is_generated, generation_expression FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"
    INDEXES_QUERY = "SELECT index_id, table_id, index_name, index_type, is_unique FROM sys.indexes WHERE is_valid = 1 ORDER BY table_id, index_name"
    INDEX_COLUMNS_QUERY = "SELECT index_id, column_id, column_name, ordinal_position, is_included FROM sys.index_columns ORDER BY index_id, ordinal_position"
    CONSTRAINTS_QUERY = "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 ORDER BY table_id, constraint_name"
    PROCEDURES_QUERY = "SELECT procedure_id, schema_id, procedure_name, routine_type FROM sys.procedures WHERE is_valid = 1 ORDER BY schema_id, procedure_name"
    FUNCTIONS_QUERY = "SELECT function_id, schema_id, function_name FROM sys.functions WHERE is_valid = 1 ORDER BY schema_id, function_name"
    CATALOGS_QUERY = "SELECT DISTINCT '' AS catalog_name FROM sys.schemas ORDER BY catalog_name"
    TYPES_QUERY = "SELECT DISTINCT data_type_id AS type_id, data_type_name AS type_name FROM sys.columns WHERE is_valid = 1 ORDER BY type_name"
    PRIMARY_KEYS_QUERY = "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 AND LOWER(constraint_type) IN ('primary key', 'primary_key', 'primary') ORDER BY table_id, constraint_name"
    FOREIGN_KEYS_QUERY = "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 AND LOWER(constraint_type) IN ('foreign key', 'foreign_key', 'foreign') ORDER BY table_id, constraint_name"
    TABLE_PRIVILEGES_QUERY = "SELECT table_id, schema_id, table_name, '' AS grantor, '' AS grantee, 'SELECT' AS privilege_type, 0 AS is_grantable FROM sys.tables WHERE is_valid = 1 ORDER BY table_name"
    COLUMN_PRIVILEGES_QUERY = "SELECT c.column_id, c.table_id, c.column_name, t.table_name, '' AS grantor, '' AS grantee, 'SELECT' AS privilege_type, 0 AS is_grantable FROM sys.columns c JOIN sys.tables t ON t.table_id = c.table_id WHERE c.is_valid = 1 AND t.is_valid = 1 ORDER BY c.table_id, c.ordinal_position"
    DDL_FIELDS_QUERY = "SELECT table_id, column_id, column_name, data_type_name, default_value, generation_expression, is_nullable, is_identity, is_generated FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"
    COLLECTION_QUERIES = {
      "schemas" => SCHEMAS_QUERY,
      "tables" => TABLES_QUERY,
      "columns" => COLUMNS_QUERY,
      "indexes" => INDEXES_QUERY,
      "index_columns" => INDEX_COLUMNS_QUERY,
      "constraints" => CONSTRAINTS_QUERY,
      "procedures" => PROCEDURES_QUERY,
      "functions" => FUNCTIONS_QUERY,
      "catalogs" => CATALOGS_QUERY,
      "types" => TYPES_QUERY,
      "primary_keys" => PRIMARY_KEYS_QUERY,
      "foreign_keys" => FOREIGN_KEYS_QUERY,
      "table_privileges" => TABLE_PRIVILEGES_QUERY,
      "column_privileges" => COLUMN_PRIVILEGES_QUERY,
      "ddl_fields" => DDL_FIELDS_QUERY
    }.freeze
    COLLECTION_ALIASES = {
      "schema" => "schemas",
      "schemas" => "schemas",
      "table" => "tables",
      "tables" => "tables",
      "column" => "columns",
      "columns" => "columns",
      "index" => "indexes",
      "indexes" => "indexes",
      "indexcolumns" => "index_columns",
      "index_columns" => "index_columns",
      "constraint" => "constraints",
      "constraints" => "constraints",
      "procedure" => "procedures",
      "procedures" => "procedures",
      "function" => "functions",
      "functions" => "functions",
      "catalog" => "catalogs",
      "catalogs" => "catalogs",
      "type" => "types",
      "types" => "types",
      "primarykey" => "primary_keys",
      "primarykeys" => "primary_keys",
      "primary_key" => "primary_keys",
      "primary_keys" => "primary_keys",
      "foreignkey" => "foreign_keys",
      "foreignkeys" => "foreign_keys",
      "foreign_key" => "foreign_keys",
      "foreign_keys" => "foreign_keys",
      "tableprivilege" => "table_privileges",
      "tableprivileges" => "table_privileges",
      "table_privilege" => "table_privileges",
      "table_privileges" => "table_privileges",
      "columnprivilege" => "column_privileges",
      "columnprivileges" => "column_privileges",
      "column_privilege" => "column_privileges",
      "column_privileges" => "column_privileges",
      "ddlfield" => "ddl_fields",
      "ddlfields" => "ddl_fields",
      "ddl_field" => "ddl_fields",
      "ddl_fields" => "ddl_fields"
    }.freeze
    SCHEMA_FIELD_CANDIDATES = %w[
      schema_name
      table_schem
      table_schema
      schema
      SCHEMA_NAME
      TABLE_SCHEM
      TABLE_SCHEMA
      SCHEMA
    ].freeze
    RESTRICTION_KEY_ALIASES = {
      "catalog" => %w[catalog_name table_catalog table_cat catalog],
      "schema" => %w[schema_name table_schema table_schem schema],
      "table" => %w[table_name table relname],
      "column" => %w[column_name column],
      "index" => %w[index_name index],
      "constraint" => %w[constraint_name constraint],
      "procedure" => %w[procedure_name routine_name],
      "function" => %w[function_name routine_name],
      "type" => %w[type_name data_type_name data_type udt_name],
      "grantor" => %w[grantor],
      "grantee" => %w[grantee],
      "privilege" => %w[privilege_type privilege],
      "ownerid" => %w[owner_id ownerid owner],
      "pkname" => %w[pk_name primary_key_name constraint_name],
      "fkname" => %w[fk_name foreign_key_name constraint_name],
      "keyname" => %w[key_name constraint_name]
    }.freeze
    COLLECTION_RESTRICTION_KEYS = {
      "schemas" => %w[catalog schema],
      "tables" => %w[catalog schema table type ownerid],
      "columns" => %w[catalog schema table column type],
      "indexes" => %w[catalog schema table index],
      "index_columns" => %w[catalog schema table index column],
      "constraints" => %w[catalog schema table constraint],
      "procedures" => %w[catalog schema procedure],
      "functions" => %w[catalog schema function],
      "catalogs" => %w[catalog],
      "types" => %w[catalog schema type],
      "primary_keys" => %w[catalog schema table key_name pk_name constraint],
      "foreign_keys" => %w[catalog schema table key_name fk_name constraint],
      "table_privileges" => %w[catalog schema table grantee grantor privilege],
      "column_privileges" => %w[catalog schema table column grantee grantor privilege],
      "ddl_fields" => %w[catalog schema table column type]
    }.freeze

    class SchemaTreeNode
      attr_reader :name, :full_path, :children
      attr_accessor :terminal

      def initialize(name, full_path)
        @name = name
        @full_path = full_path
        @terminal = false
        @children = []
      end
    end

    def self.schemas_query
      SCHEMAS_QUERY
    end

    def self.tables_query
      TABLES_QUERY
    end

    def self.columns_query
      COLUMNS_QUERY
    end

    def self.indexes_query
      INDEXES_QUERY
    end

    def self.index_columns_query
      INDEX_COLUMNS_QUERY
    end

    def self.constraints_query
      CONSTRAINTS_QUERY
    end

    def self.procedures_query
      PROCEDURES_QUERY
    end

    def self.functions_query
      FUNCTIONS_QUERY
    end

    def self.normalize_collection_name(collection_name = "tables")
      raw = collection_name.to_s.strip.downcase
      raw = "tables" if raw.empty?
      collapsed = raw.gsub(/[^a-z0-9]/, "")
      normalized = COLLECTION_ALIASES[raw] || COLLECTION_ALIASES[collapsed]
      return normalized if normalized

      raise ArgumentError, "Metadata collection '#{collection_name}' is not supported"
    end

    def self.resolve_collection_query(collection_name = "tables")
      COLLECTION_QUERIES.fetch(normalize_collection_name(collection_name))
    end

    def self.normalize_restrictions(restrictions)
      return {} if restrictions.nil?
      return {} if restrictions.respond_to?(:empty?) && restrictions.empty?
      unless restrictions.respond_to?(:each_pair)
        raise ArgumentError, "metadata restrictions must be provided as a hash"
      end

      out = {}
      restrictions.each_pair do |key, value|
        normalized = normalize_identifier(key)
        next if normalized.empty?

        out[normalized] = value
      end
      out
    end

    def self.filter_rows_by_restrictions(rows, restrictions, collection_name: nil)
      normalized_restrictions = normalize_restrictions(restrictions)
      return rows if normalized_restrictions.empty?

      bindings = build_restriction_bindings(rows, normalized_restrictions, collection_name)
      return rows if bindings.empty?

      rows.select { |row| row_matches_restrictions?(row, bindings) }
    end

    def self.schema_paths_for_navigation(rows_or_names, expand_schema_parents: false)
      base_paths = unique_schema_paths(rows_or_names)
      return base_paths unless expand_schema_parents

      expand_schema_parent_paths(base_paths)
    end

    def self.expand_schema_parent_paths(rows_or_names)
      out = []
      seen = {}

      each_schema_path(rows_or_names) do |schema_path|
        current = +""
        split_schema_path(schema_path).each do |segment|
          current = current.empty? ? segment : "#{current}.#{segment}"
          next if seen[current]

          seen[current] = true
          out << current
        end
      end

      out
    end

    def self.build_schema_tree(schema_paths)
      nodes_by_path = {}
      roots = []

      each_schema_path(schema_paths) do |schema_path|
        parent = nil
        path_parts = []

        split_schema_path(schema_path).each do |segment|
          path_parts << segment
          full_path = path_parts.join(".")

          node = nodes_by_path[full_path]
          unless node
            node = SchemaTreeNode.new(segment, full_path)
            nodes_by_path[full_path] = node
            if parent
              parent.children << node
            else
              roots << node
            end
          end

          parent = node
        end

        parent.terminal = true if parent
      end

      roots
    end

    def self.expand_schema_metadata_rows(rows)
      out = []
      seen = {}

      rows.each do |row|
        schema_path = read_schema_path(row)
        unless schema_path
          out << row
          next
        end

        current = +""
        segments = split_schema_path(schema_path)
        segments.each_with_index do |segment, idx|
          current = current.empty? ? segment : "#{current}.#{segment}"
          next if seen[current]

          seen[current] = true
          if idx == segments.length - 1
            out << row
          else
            out << synthetic_schema_row(row, current)
          end
        end
      end

      out
    end

    def self.build_database_default_metadata_rows(rows_or_names, database:, expand_schema_parents: false, default_branch: "default")
      database_name = database.to_s.strip
      raise ArgumentError, "database is required" if database_name.empty?

      default_name = default_branch.to_s.strip
      raise ArgumentError, "default_branch is required" if default_name.empty?

      schema_paths = schema_paths_for_navigation(rows_or_names, expand_schema_parents: expand_schema_parents)
      roots = build_schema_tree(schema_paths)
      rows = []

      database_path = database_name
      default_path = "#{database_path}.#{default_name}"

      rows << {
        "node_type" => "database",
        "node_name" => database_name,
        "node_path" => database_path,
        "parent_path" => nil,
        "schema_path" => nil,
        "terminal" => false
      }

      rows << {
        "node_type" => "schema",
        "node_name" => default_name,
        "node_path" => default_path,
        "parent_path" => database_path,
        "schema_path" => nil,
        "terminal" => false
      }

      append_tree_metadata_rows(rows, roots, default_path)
      rows
    end

    def self.append_tree_metadata_rows(out_rows, nodes, parent_node_path)
      nodes.each do |node|
        node_path = "#{parent_node_path}.#{node.name}"
        out_rows << {
          "node_type" => "schema",
          "node_name" => node.name,
          "node_path" => node_path,
          "parent_path" => parent_node_path,
          "schema_path" => node.full_path,
          "terminal" => node.terminal
        }
        append_tree_metadata_rows(out_rows, node.children, node_path)
      end
    end
    private_class_method :append_tree_metadata_rows

    def self.unique_schema_paths(rows_or_names)
      out = []
      seen = {}

      each_schema_path(rows_or_names) do |schema_path|
        next if seen[schema_path]

        seen[schema_path] = true
        out << schema_path
      end

      out
    end
    private_class_method :unique_schema_paths

    def self.each_schema_path(rows_or_names)
      return enum_for(:each_schema_path, rows_or_names) unless block_given?
      return if rows_or_names.nil?

      enumerated =
        if rows_or_names.is_a?(String) || rows_or_names.is_a?(Hash)
          [rows_or_names]
        elsif rows_or_names.respond_to?(:each)
          rows_or_names
        else
          [rows_or_names]
        end

      enumerated.each do |row_or_name|
        schema_path = read_schema_path(row_or_name)
        next unless schema_path

        yield schema_path
      end
    end
    private_class_method :each_schema_path

    def self.read_schema_path(row_or_name)
      return normalize_schema_path(row_or_name) if row_or_name.is_a?(String)

      return normalize_schema_path(row_or_name.to_s) if row_or_name.is_a?(Symbol)

      return read_schema_path_from_hash(row_or_name) if row_or_name.is_a?(Hash)

      nil
    end
    private_class_method :read_schema_path

    def self.read_schema_path_from_hash(row)
      SCHEMA_FIELD_CANDIDATES.each do |candidate|
        value = nil
        value = row[candidate] if row.key?(candidate)
        symbol = candidate.to_sym
        value = row[symbol] if value.nil? && row.key?(symbol)
        next unless value.is_a?(String)

        normalized = normalize_schema_path(value)
        return normalized if normalized
      end

      nil
    end
    private_class_method :read_schema_path_from_hash

    def self.normalize_schema_path(value)
      normalized = split_schema_path(value).join(".")
      return nil if normalized.empty?

      normalized
    end
    private_class_method :normalize_schema_path

    def self.split_schema_path(value)
      value.to_s.split(".").map(&:strip).reject(&:empty?)
    end
    private_class_method :split_schema_path

    def self.synthetic_schema_row(sample_row, schema_path)
      synthetic = {}
      sample_row.each_key do |key|
        synthetic[key] = nil
      end
      assign_schema_path(synthetic, schema_path)
      synthetic
    end
    private_class_method :synthetic_schema_row

    def self.assign_schema_path(row, schema_path)
      assigned = false
      SCHEMA_FIELD_CANDIDATES.each do |candidate|
        if row.key?(candidate)
          row[candidate] = schema_path
          assigned = true
        end
        symbol = candidate.to_sym
        if row.key?(symbol)
          row[symbol] = schema_path
          assigned = true
        end
      end
      return if assigned

      target_key = row.keys.any? { |key| key.is_a?(Symbol) } ? :schema_name : "schema_name"
      row[target_key] = schema_path
    end
    private_class_method :assign_schema_path

    def self.build_restriction_bindings(rows, normalized_restrictions, collection_name)
      aliases_allowed =
        if collection_name
          normalized_collection = normalize_collection_name(collection_name)
          Array(COLLECTION_RESTRICTION_KEYS[normalized_collection]).flat_map do |key|
            metadata_restriction_key_aliases(key)
          end.map { |alias_key| normalize_identifier(alias_key) }.reject(&:empty?).uniq
        else
          []
        end

      bindings = normalized_restrictions.map do |key, value|
        normalized_key = normalize_identifier(key)
        if !aliases_allowed.empty? && !aliases_allowed.include?(normalized_key)
          next
        end
        aliases = metadata_restriction_key_aliases(key)
                  .map { |alias_key| normalize_identifier(alias_key) }
                  .reject(&:empty?)
                  .uniq
        aliases &= aliases_allowed unless aliases_allowed.empty?
        aliases |= [normalized_key] if aliases_allowed.empty? || aliases_allowed.include?(normalized_key)
        next if aliases.empty?

        expected_text = normalize_match_text(value)
        expect_null = expected_text == "null"
        {
          aliases: aliases,
          expect_null: expect_null,
          expected_text: expect_null ? nil : expected_text
        }
      end.compact

      bindings.select do |binding|
        rows_have_binding_aliases?(rows: rows, binding: binding)
      end
    end
    private_class_method :build_restriction_bindings

    def self.row_matches_restrictions?(row, bindings)
      bindings.all? do |binding|
        values = metadata_values_for_aliases(row, binding[:aliases])
        next false if values.empty?

        values.any? do |value|
          if binding[:expect_null]
            value.nil?
          else
            !value.nil? && normalize_match_text(value) == binding[:expected_text]
          end
        end
      end
    end
    private_class_method :row_matches_restrictions?

    def self.metadata_restriction_key_aliases(key)
      canonical = normalize_identifier(key)
      aliases = Array(RESTRICTION_KEY_ALIASES[canonical])
      aliases << canonical unless canonical.empty?
      aliases
    end
    private_class_method :metadata_restriction_key_aliases

    def self.metadata_values_for_aliases(row, aliases)
      return [] unless row.is_a?(Hash)

      values = []
      aliases.each do |alias_key|
        present, value = metadata_row_value(row, alias_key)
        values << value if present
      end
      values
    end
    private_class_method :metadata_values_for_aliases

    def self.metadata_row_value(row, alias_key)
      return [true, row[alias_key]] if row.key?(alias_key)

      symbol = alias_key.to_sym
      return [true, row[symbol]] if row.key?(symbol)

      target = normalize_identifier(alias_key)
      row.each do |candidate, value|
        return [true, value] if normalize_identifier(candidate) == target
      end
      [false, nil]
    end
    private_class_method :metadata_row_value

    def self.rows_have_binding_aliases?(rows:, binding:)
      return false unless rows.respond_to?(:any?)

      rows.any? do |row|
        binding[:aliases].any? { |alias_key| metadata_row_value(row, alias_key).first }
      end
    end
    private_class_method :rows_have_binding_aliases?

    def self.normalize_identifier(value)
      value.to_s.strip.downcase.gsub(/[^a-z0-9]/, "")
    end
    private_class_method :normalize_identifier

    def self.normalize_match_text(value)
      value.to_s.strip.downcase
    end
    private_class_method :normalize_match_text
  end
end

File: project/drivers/driver/swift/Sources/ScratchBird/Metadata.swift
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import Foundation

public enum ScratchBirdMetadata {
    public static let schemasQuery =
        "SELECT schema_id, schema_name, owner_id, default_tablespace_id FROM sys.schemas WHERE is_valid = 1 ORDER BY schema_name"

    public static let tablesQuery =
        "SELECT table_id, schema_id, table_name, table_type, owner_id FROM sys.tables WHERE is_valid = 1 ORDER BY table_name"

    public static let columnsQuery =
        "SELECT column_id, table_id, column_name, data_type_id, data_type_name, ordinal_position, is_nullable, default_value, domain_id, collation_id, charset_id, is_identity, is_generated, generation_expression FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"

    public static let indexesQuery =
        "SELECT index_id, table_id, index_name, index_type, is_unique FROM sys.indexes WHERE is_valid = 1 ORDER BY table_id, index_name"

    public static let indexColumnsQuery =
        "SELECT index_id, column_id, column_name, ordinal_position, is_included FROM sys.index_columns ORDER BY index_id, ordinal_position"

    public static let constraintsQuery =
        "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 ORDER BY table_id, constraint_name"

    public static let proceduresQuery =
        "SELECT routine_schema AS schema_id, routine_schema AS schema_name, routine_schema AS table_schema, routine_schema AS table_schem, routine_name AS procedure_name, routine_name AS routine_name, routine_name AS specific_name, routine_type FROM information_schema.routines WHERE lower(routine_type) = 'procedure' ORDER BY schema_name, procedure_name"

    public static let functionsQuery =
        "SELECT routine_schema AS schema_id, routine_schema AS schema_name, routine_schema AS table_schema, routine_schema AS table_schem, routine_name AS function_name, routine_name AS routine_name, routine_name AS specific_name, routine_type FROM information_schema.routines WHERE lower(routine_type) = 'function' ORDER BY schema_name, function_name"

    public static let routinesQuery =
        "SELECT routine_schema AS schema_id, routine_schema AS schema_name, routine_schema AS table_schema, routine_schema AS table_schem, routine_name AS routine_name, routine_name AS specific_name, routine_type FROM information_schema.routines ORDER BY schema_name, routine_name"

    public static let catalogsQuery =
        "SELECT schema_id AS catalog_id, schema_name AS catalog_name FROM sys.schemas WHERE is_valid = 1 ORDER BY schema_name"

    public static let primaryKeysQuery =
        "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 AND lower(constraint_type) IN ('primary key', 'primary') ORDER BY table_id, constraint_name"

    public static let foreignKeysQuery =
        "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 AND lower(constraint_type) IN ('foreign key', 'foreign') ORDER BY table_id, constraint_name"

    public static let tablePrivilegesQuery =
        "SELECT table_id, table_name, owner_id AS grantor_id, owner_id AS grantee_id, 'ALL' AS privilege_type FROM sys.tables WHERE is_valid = 1 ORDER BY table_id, table_name"

    public static let columnPrivilegesQuery =
        "SELECT table_id, column_id, column_name, 'ALL' AS privilege_type FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"

    public static let typeInfoQuery =
        "SELECT DISTINCT data_type_id, data_type_name, data_type_name AS type_name FROM sys.columns WHERE is_valid = 1 ORDER BY data_type_name"
}

public enum ScratchBirdMetadataCollectionName: String {
    case schemas
    case tables
    case columns
    case indexes
    case indexColumns = "index_columns"
    case constraints
    case procedures
    case functions
    case routines
    case catalogs
    case primaryKeys = "primary_keys"
    case foreignKeys = "foreign_keys"
    case tablePrivileges = "table_privileges"
    case columnPrivileges = "column_privileges"
    case typeInfo = "type_info"
}

private let metadataCollectionQueries: [ScratchBirdMetadataCollectionName: String] = [
    .schemas: ScratchBirdMetadata.schemasQuery,
    .tables: ScratchBirdMetadata.tablesQuery,
    .columns: ScratchBirdMetadata.columnsQuery,
    .indexes: ScratchBirdMetadata.indexesQuery,
    .indexColumns: ScratchBirdMetadata.indexColumnsQuery,
    .constraints: ScratchBirdMetadata.constraintsQuery,
    .procedures: ScratchBirdMetadata.proceduresQuery,
    .functions: ScratchBirdMetadata.functionsQuery,
    .routines: ScratchBirdMetadata.routinesQuery,
    .catalogs: ScratchBirdMetadata.catalogsQuery,
    .primaryKeys: ScratchBirdMetadata.primaryKeysQuery,
    .foreignKeys: ScratchBirdMetadata.foreignKeysQuery,
    .tablePrivileges: ScratchBirdMetadata.tablePrivilegesQuery,
    .columnPrivileges: ScratchBirdMetadata.columnPrivilegesQuery,
    .typeInfo: ScratchBirdMetadata.typeInfoQuery,
]

private let metadataCollectionAliases: [String: ScratchBirdMetadataCollectionName] = [
    "schema": .schemas,
    "schemas": .schemas,
    "table": .tables,
    "tables": .tables,
    "column": .columns,
    "columns": .columns,
    "index": .indexes,
    "indexes": .indexes,
    "indexcolumn": .indexColumns,
    "indexcolumns": .indexColumns,
    "index_column": .indexColumns,
    "index_columns": .indexColumns,
    "constraint": .constraints,
    "constraints": .constraints,
    "procedure": .procedures,
    "procedures": .procedures,
    "function": .functions,
    "functions": .functions,
    "routine": .routines,
    "routines": .routines,
    "catalog": .catalogs,
    "catalogs": .catalogs,
    "primary_key": .primaryKeys,
    "primary_keys": .primaryKeys,
    "primarykey": .primaryKeys,
    "primarykeys": .primaryKeys,
    "pk": .primaryKeys,
    "foreign_key": .foreignKeys,
    "foreign_keys": .foreignKeys,
    "foreignkey": .foreignKeys,
    "foreignkeys": .foreignKeys,
    "fk": .foreignKeys,
    "table_privilege": .tablePrivileges,
    "table_privileges": .tablePrivileges,
    "tableprivilege": .tablePrivileges,
    "tableprivileges": .tablePrivileges,
    "column_privilege": .columnPrivileges,
    "column_privileges": .columnPrivileges,
    "columnprivilege": .columnPrivileges,
    "columnprivileges": .columnPrivileges,
    "type_info": .typeInfo,
    "typeinfo": .typeInfo,
    "types": .typeInfo,
]

public func normalizeMetadataCollectionName(_ collectionName: String? = nil) throws -> ScratchBirdMetadataCollectionName {
    let raw = (collectionName ?? "tables").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    if let resolved = metadataCollectionAliases[raw] {
        return resolved
    }
    throw ScratchBirdNotSupportedException(message: "metadata collection '\(collectionName ?? "")' is not supported", sqlState: "0A000")
}

public func resolveMetadataCollectionQuery(_ collectionName: String? = nil) throws -> String {
    let resolved = try normalizeMetadataCollectionName(collectionName)
    return metadataCollectionQueries[resolved]!
}

public func metadataSchemaNames(from result: ScratchBirdResult) -> [String] {
    let index = metadataSchemaNameColumnIndex(result.columns) ?? inferredSchemaNameColumnIndex(result.rows)
    guard let index else { return [] }

    var out: [String] = []
    var seen: Set<String> = []
    for row in result.rows {
        if index >= row.count {
            continue
        }
        guard let value = metadataStringValue(row[index]) else {
            continue
        }
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if normalized.isEmpty {
            continue
        }
        if seen.insert(normalized).inserted {
            out.append(normalized)
        }
    }
    return out
}

public enum ScratchBirdMetadataTreeRowKind: Equatable {
    case database
    case schema
}

public struct ScratchBirdMetadataSchemaTreeRow {
    public let kind: ScratchBirdMetadataTreeRowKind
    public let database: String
    public let parentPath: String
    public let path: String
    public let name: String
    public let terminal: Bool
    public let topLevelBranch: Bool

    public init(
        kind: ScratchBirdMetadataTreeRowKind,
        database: String,
        parentPath: String,
        path: String,
        name: String,
        terminal: Bool,
        topLevelBranch: Bool
    ) {
        self.kind = kind
        self.database = database
        self.parentPath = parentPath
        self.path = path
        self.name = name
        self.terminal = terminal
        self.topLevelBranch = topLevelBranch
    }
}

public final class ScratchBirdMetadataSchemaTreeNode {
    public let name: String
    public let path: String
    public var terminal: Bool
    public var children: [ScratchBirdMetadataSchemaTreeNode]

    public init(
        name: String,
        path: String,
        terminal: Bool = false,
        children: [ScratchBirdMetadataSchemaTreeNode] = []
    ) {
        self.name = name
        self.path = path
        self.terminal = terminal
        self.children = children
    }
}

public struct ScratchBirdMetadataSchemaTree {
    public let database: String?
    public let schemas: [ScratchBirdMetadataSchemaTreeNode]

    public init(database: String?, schemas: [ScratchBirdMetadataSchemaTreeNode]) {
        self.database = database
        self.schemas = schemas
    }
}

public func metadataSchemaPathsForNavigation(
    _ schemaNames: [String],
    expandSchemaParents: Bool = false
) -> [String] {
    var out: [String] = []
    var seen: Set<String> = []

    for raw in schemaNames {
        guard let normalized = normalizeMetadataSchemaPath(raw) else {
            continue
        }

        if !expandSchemaParents {
            if seen.insert(normalized).inserted {
                out.append(normalized)
            }
            continue
        }

        var current = ""
        for part in splitMetadataSchemaPath(normalized) {
            current = current.isEmpty ? part : "\(current).\(part)"
            if seen.insert(current).inserted {
                out.append(current)
            }
        }
    }

    return out
}

public func buildMetadataSchemaTree(
    _ schemaNames: [String],
    database: String = "",
    expandSchemaParents: Bool = false
) -> ScratchBirdMetadataSchemaTree {
    let schemaPaths = metadataSchemaPathsForNavigation(schemaNames, expandSchemaParents: expandSchemaParents)
    let terminalPaths = Set(schemaPaths)
    var nodesByPath: [String: ScratchBirdMetadataSchemaTreeNode] = [:]
    var roots: [ScratchBirdMetadataSchemaTreeNode] = []

    for schemaPath in schemaPaths {
        var parent: ScratchBirdMetadataSchemaTreeNode?
        var currentPath = ""
        for part in splitMetadataSchemaPath(schemaPath) {
            currentPath = currentPath.isEmpty ? part : "\(currentPath).\(part)"
            let node: ScratchBirdMetadataSchemaTreeNode
            if let existing = nodesByPath[currentPath] {
                node = existing
            } else {
                node = ScratchBirdMetadataSchemaTreeNode(name: part, path: currentPath)
                nodesByPath[currentPath] = node
                if let parent {
                    parent.children.append(node)
                } else {
                    roots.append(node)
                }
            }
            if terminalPaths.contains(currentPath) {
                node.terminal = true
            }
            parent = node
        }
    }

    let trimmedDatabase = database.trimmingCharacters(in: .whitespacesAndNewlines)
    let normalizedDatabase = trimmedDatabase.isEmpty ? nil : trimmedDatabase
    return ScratchBirdMetadataSchemaTree(database: normalizedDatabase, schemas: roots)
}

public func buildMetadataSchemaTreeRows(
    _ schemaNames: [String],
    database: String = "",
    expandSchemaParents: Bool = false
) -> [ScratchBirdMetadataSchemaTreeRow] {
    let tree = buildMetadataSchemaTree(
        schemaNames,
        database: database,
        expandSchemaParents: expandSchemaParents
    )
    let databaseName = tree.database ?? "default"
    var rows: [ScratchBirdMetadataSchemaTreeRow] = [
        ScratchBirdMetadataSchemaTreeRow(
            kind: .database,
            database: databaseName,
            parentPath: "",
            path: databaseName,
            name: databaseName,
            terminal: false,
            topLevelBranch: false
        )
    ]

    for root in tree.schemas {
        appendMetadataTreeRowsDepthFirst(
            root,
            databaseName: databaseName,
            parentPath: databaseName,
            topLevel: true,
            rows: &rows
        )
    }

    return rows
}

public func splitMetadataSchemaPath(_ schemaName: String) -> [String] {
    schemaName
        .split(separator: ".", omittingEmptySubsequences: false)
        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
}

public func normalizeMetadataSchemaPath(_ schemaName: String) -> String? {
    let normalized = splitMetadataSchemaPath(schemaName).joined(separator: ".")
    return normalized.isEmpty ? nil : normalized
}

private func appendMetadataTreeRowsDepthFirst(
    _ node: ScratchBirdMetadataSchemaTreeNode,
    databaseName: String,
    parentPath: String,
    topLevel: Bool,
    rows: inout [ScratchBirdMetadataSchemaTreeRow]
) {
    rows.append(
        ScratchBirdMetadataSchemaTreeRow(
            kind: .schema,
            database: databaseName,
            parentPath: parentPath,
            path: node.path,
            name: node.name,
            terminal: node.terminal,
            topLevelBranch: topLevel
        )
    )
    for child in node.children {
        appendMetadataTreeRowsDepthFirst(
            child,
            databaseName: databaseName,
            parentPath: node.path,
            topLevel: false,
            rows: &rows
        )
    }
}

private func metadataSchemaNameColumnIndex(_ columns: [ScratchBirdColumn]) -> Int? {
    for (index, column) in columns.enumerated() {
        let lowered = column.name.lowercased()
        if lowered == "schema_name" || lowered == "nspname" || lowered == "schema" {
            return index
        }
    }
    return nil
}

private func inferredSchemaNameColumnIndex(_ rows: [[Any?]]) -> Int? {
    guard let first = rows.first else { return nil }
    if first.count > 1 {
        return 1
    }
    if first.count == 1 {
        return 0
    }
    return nil
}

private func metadataStringValue(_ value: Any?) -> String? {
    guard let value else { return nil }
    if let text = value as? String {
        return text
    }
    if let raw = value as? RawValue {
        return String(data: raw.data, encoding: .utf8)
    }
    if let data = value as? Data {
        return String(data: data, encoding: .utf8)
    }
    if let value = value as? CustomStringConvertible {
        return value.description
    }
    return nil
}
